data/transactions/logic/resources.go

```diff
@@ -0,0 +1,359 @@
+	// the box counts against the group writeBudget. So delete is NOT a dirtying
+	// operation.
+	boxes map[boxRef]bool
+
+	// dirtyBytes maintains a running count of the number of dirty bytes in `boxes`
+	dirtyBytes uint64
+}
+
+func (r *resources) shareHolding(addr basics.Address, id basics.AssetIndex) {
+	r.sharedHoldings[ledgercore.AccountAsset{Address: addr, Asset: id}] = struct{}{}
+}
+
+func (r *resources) shareAccountAndHolding(addr basics.Address, id basics.AssetIndex) {
+	r.sharedAccounts[addr] = struct{}{}
+	if id != 0 {
+		r.sharedHoldings[ledgercore.AccountAsset{Address: addr, Asset: id}] = struct{}{}
+	}
+}
+
+func (r *resources) shareLocal(addr basics.Address, id basics.AppIndex) {
+	r.sharedLocals[ledgercore.AccountApp{Address: addr, App: id}] = struct{}{}
+}
+
+// In the fill* and allows* routines, we pass the header and the fields in
+// separately, even though they are pointers into the same structure. That
+// prevents dumb attempts to use other fields from the transaction.
+
+func (r *resources) fill(tx *transactions.Transaction, ep *EvalParams) {
+	switch tx.Type {
+	case protocol.PaymentTx:
+		r.fillPayment(&tx.Header, &tx.PaymentTxnFields)
+	case protocol.KeyRegistrationTx:
+		r.fillKeyRegistration(&tx.Header)
+	case protocol.AssetConfigTx:
+		r.fillAssetConfig(&tx.Header, &tx.AssetConfigTxnFields)
+	case protocol.AssetTransferTx:
+		r.fillAssetTransfer(&tx.Header, &tx.AssetTransferTxnFields)
+	case protocol.AssetFreezeTx:
+		r.fillAssetFreeze(&tx.Header, &tx.AssetFreezeTxnFields)
+	case protocol.ApplicationCallTx:
+		r.fillApplicationCall(ep, &tx.Header, &tx.ApplicationCallTxnFields)
+	case protocol.StateProofTx:
+		// state proof txns add nothing to availability (they can't even appear
+		// in a group with an appl. but still.)
+	default:
+		panic(tx.Type)
+	}
+}
+
+func (cx *EvalContext) allows(tx *transactions.Transaction, calleeVer uint64) error {
+	switch tx.Type {
+	case protocol.PaymentTx, protocol.KeyRegistrationTx, protocol.AssetConfigTx:
+		// these transactions don't touch cross-product resources, so no error is possible
+		return nil
+	case protocol.AssetTransferTx:
+		return cx.allowsAssetTransfer(&tx.Header, &tx.AssetTransferTxnFields)
+	case protocol.AssetFreezeTx:
+		return cx.allowsAssetFreeze(&tx.Header, &tx.AssetFreezeTxnFields)
+	case protocol.ApplicationCallTx:
+		return cx.allowsApplicationCall(&tx.Header, &tx.ApplicationCallTxnFields, cx.version, calleeVer)
+	default:
+		return fmt.Errorf("unknown inner transaction type %s", tx.Type)
+	}
+}
+
+func (r *resources) fillKeyRegistration(hdr *transactions.Header) {
+	r.sharedAccounts[hdr.Sender] = struct{}{}
+}
+
+func (r *resources) fillPayment(hdr *transactions.Header, tx *transactions.PaymentTxnFields) {
+	r.sharedAccounts[hdr.Sender] = struct{}{}
+	r.sharedAccounts[tx.Receiver] = struct{}{}
+	if !tx.CloseRemainderTo.IsZero() {
+		r.sharedAccounts[tx.CloseRemainderTo] = struct{}{}
+	}
+}
+
```

Comment:
can you elaborate on "special addresses"

Reply:
AssetConfig takes addresses, the Manager, Freezer, Reserve, and Clawback.

But these addresses are used like the Rekey field. Although they seem like accounts, they need not exist at the time of acfg, and are not looked up. So they should not be added to availability.

Do you think this merits being put in as a comment?

Reply:
up to you